tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla

```diff
@@ -0,0 +1,384 @@
+
+--------------------------- MODULE RandomAccessFile ---------------------------
+
+EXTENDS Naturals, Sequences, Common
+
+VARIABLES
+    file_content,
+    file_pointer
+
+vars == <<file_content, file_pointer>>
+
+TypeOK ==
+    /\ file_content \in ArrayOfAnyLength(Symbols \union {ArbitrarySymbol})
+    /\ ArrayLen(file_content) <= MaxOffset
+    /\ file_pointer \in Offset
+
+Init ==
+    /\ file_content = EmptyArray
+    /\ file_pointer = 0
+
+Seek(new_offset) ==
+    /\ new_offset \in Offset
+    /\ file_pointer' = new_offset
+    /\ UNCHANGED <<file_content>>
+
+SetLength(new_length) ==
+    /\ file_content' = TruncateOrExtendFile(file_content, new_length)
+    /\ file_pointer' = Min(file_pointer, new_length)
+
+Read(output) ==
+    /\ output = ArraySlice(file_content, file_pointer, Min(file_pointer + ArrayLen(output), ArrayLen(file_content)))
+    /\ file_pointer' = file_pointer + ArrayLen(output)
+    /\ UNCHANGED <<file_content>>
+
+Write(data) ==
+    /\ file_pointer + ArrayLen(data) <= MaxOffset
+    /\ file_content' = WriteToFile(file_content, file_pointer, data)
+    /\ file_pointer' = file_pointer + ArrayLen(data)
+
+Next ==
+    \/ \E offset \in Offset:
+        \/ Seek(offset)
+        \/ SetLength(offset)
+    \/ \E len \in 1..MaxOffset: \E data \in Array(Symbols, len):
+        \/ Write(data)
+        \/ Read(data)
+
+Spec == Init /\ [][Next]_vars
+
+===============================================================================
+
+-------------------------------- MODULE Common --------------------------------
+
+EXTENDS Naturals, Sequences
+
+CONSTANTS
+    Symbols,
+    ArbitrarySymbol,
+    MaxOffset
+
+Offset == 0..MaxOffset
+
+Min(a, b) == IF a <= b THEN a ELSE b
+Max(a, b) == IF a <= b THEN b ELSE a
+
+\* 0-indexed arrays
+ArrayOfAnyLength(T) == [elems: Seq(T)] \* could use Seq(T), but we don't want a[i] to work; have to use ArrayGet
+Array(T, len) == [elems: [1..len -> T]]
+MkArray(len, f) == [elems |-> [i \in 1..len |-> f[i - 1]]]
+EmptyArray == [elems |-> <<>>]
+ArrayLen(a) == Len(a.elems)
+ArrayToSeq(a) == a.elems
+SeqToArray(seq) == [elems |-> seq]
+ArrayGet(a, i) == a.elems[i+1]
+ArraySet(a, i, x) == [a EXCEPT !.elems[i+1] = x]
+ArraySlice(a, startInclusive, endExclusive) == [elems |-> SubSeq(a.elems, startInclusive + 1, endExclusive)]
+ArrayConcat(a1, a2) == [elems |-> a1.elems \o a2.elems]
+
+WriteToFile(file, offset, data_to_write) ==
+    LET file_len == ArrayLen(file) IN
+    LET data_len == ArrayLen(data_to_write) IN
+    LET length == Max(file_len, offset + data_len) IN
+    MkArray(
+        length,
+        [i \in 0..(length-1) |->
+            CASE
+                i < offset -> IF i < file_len THEN ArrayGet(file, i) ELSE ArbitrarySymbol
+                []
+                i >= offset /\ i < offset + data_len -> ArrayGet(data_to_write, i - offset)
+                []
+                i >= offset + data_len -> IF i < file_len THEN ArrayGet(file, i) ELSE ArbitrarySymbol])
```

Comment:
i think it is invariant that the final `ELSE` never matches (unless I've missed something).  The only situation in which is greater than or equal to `offset + data_len` is when the original file extends beyond the total overlap of the new data.

I think it makes sense to avoid the `IF`, as otherwise a cursory glance makes it appear that we can end up `ArbitrarySymbol`s at the end of the file, which isn't the case.

Reply:
Yes, great catch! The whole `IF` can be simplified to its then-branch `ArrayGet(file, i)`.